kotlinx-coroutines-core/jvm/src/scheduling/CoroutineScheduler.kt

```diff
@@ -730,23 +705,17 @@ internal class CoroutineScheduler(
         // Counterpart to "tryUnpark"
         private fun tryPark() {
             if (!inStack()) {
-                parkingState.value = PARKING_ALLOWED
                 parkedWorkersStackPush(this)
                 return
-
             }
             assert { localQueue.size == 0 }
-            // Failed to get a parking permit => we are not in the stack
             while (inStack()) {
                 if (isTerminated || state == WorkerState.TERMINATED) break
-                if (parkingState.value != PARKED && !parkingState.compareAndSet(PARKING_ALLOWED, PARKED)) {
-                    return
-                }
                 tryReleaseCpu(WorkerState.PARKING)
+                workerCtl.value = PARKED
+                if (!inStack()) return
```

Comment:
This `if (!inStack()) return` works as "park avoidance", but the window between the previous check for `inStack` and this check is too short to be practically working. I'd suggest removing this line to simplify code.